example/sparse_advection/parthenon_app_inputs.cpp

```diff
@@ -0,0 +1,182 @@
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <sstream>
+#include <string>
+
+#include <parthenon/package.hpp>
+
+#include "config.hpp"
+#include "defs.hpp"
+#include "globals.hpp"
+#include "interface/variable.hpp"
+#include "sparse_advection_driver.hpp"
+#include "sparse_advection_package.hpp"
+#include "utils/error_checking.hpp"
+
+using namespace parthenon::package::prelude;
+using namespace parthenon;
+
+// *************************************************//
+// redefine some weakly linked parthenon functions *//
+// *************************************************//
+
+using sparse_advection_package::NUM_FIELDS;
+using sparse_advection_package::RealArr_t;
+
+namespace sparse_advection_example {
+
+void ProblemGenerator(MeshBlock *pmb, ParameterInput *pin) {
+  using parthenon::MetadataFlag;
+
+  auto cellbounds = pmb->cellbounds;
+  IndexRange ib = cellbounds.GetBoundsI(IndexDomain::interior);
+  IndexRange jb = cellbounds.GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = cellbounds.GetBoundsK(IndexDomain::interior);
+
+  auto coords = pmb->coords;
+  auto &data = pmb->meshblock_data.Get();
+  auto pkg = pmb->packages.Get("sparse_advection_package");
+  const auto r = pkg->Param<Real>("init_size");
+  const auto size = r * r;
+
+  const auto &x0s = pkg->Param<RealArr_t>("x0");
+  const auto &y0s = pkg->Param<RealArr_t>("y0");
+
+  for (int f = 0; f < NUM_FIELDS; ++f) {
+    // allocate the sparse field on the blocks where we get non-zero values
+    bool any_nonzero = false;
+    const Real x0 = x0s[f];
+    const Real y0 = y0s[f];
+
+    for (int k = kb.s; k <= kb.e; k++) {
+      for (int j = jb.s; j <= jb.e; j++) {
+        for (int i = ib.s; i <= ib.e; i++) {
+          auto x = coords.x1v(i) - x0;
+          auto y = coords.x2v(j) - y0;
+          auto z = coords.x3v(k);
+          auto r2 = x * x + y * y + z * z;
+          if (r2 < size) {
+            any_nonzero = true;
+            break;
+          }
+        }
+        if (any_nonzero) break;
+      }
+      if (any_nonzero) break;
+    }
```

Comment:
It really doesn't matter for this example... but I imagine the right way to do this kind of `any_nonzero` loop is as a device-side `par_reduce`, which might be a lot faster for large meshblocks.

Reply:
`par_reduce` will return the `any_nonzero` var to the host as a reduction after the kernel finishes. I think it'll be basically the same amount of code.